src/p11_child/p11_child_openssl.c

```diff
@@ -1979,198 +2187,13 @@ errno_t do_card(TALLOC_CTX *mem_ctx, struct p11_ctx *p11_ctx,
-                && strcmp(key_id_in, item->id) == 0
-                && (label_in == NULL
-                    || (label_in != NULL && item->label != NULL
-                        && strcmp(label_in, item->label) == 0))
-                && strcmp(token_name_in, token_name) == 0
-                && strcmp(module_name_in, module_file_name) == 0)
-            || (mode == OP_PREAUTH
-                && (module_name_in == NULL
-                    || (module_name_in != NULL
-                        && strcmp(module_name_in, module_file_name) == 0))
-                && (token_name_in == NULL
-                    || (token_name_in != NULL
-                        && strcmp(token_name_in, token_name) == 0))
-                && (key_id_in == NULL
-                    || (key_id_in != NULL && item->id != NULL
-                        && strcmp(key_id_in, item->id) == 0)))) {
-
-            item->uri = get_pkcs11_uri(mem_ctx, &module_info, &info, slot_id,
-                                       &token_info,
-                                       &item->attributes[1] /* label */,
-                                       &item->attributes[0] /* id */);
-            DEBUG(SSSDBG_TRACE_ALL, "uri: %s.\n", item->uri);
-
-        } else {
-            DLIST_REMOVE(cert_list, item);
-            talloc_free(item);
-        }
-    }
-
-    /* TODO: check module_name_in, token_name_in, key_id_in */
-
-    if (cert_list == NULL) {
-        DEBUG(SSSDBG_TRACE_ALL, "No certificate found.\n");
-        *_multi = NULL;
-        ret = EOK;
-        goto done;
-    }
-
-    if (mode == OP_AUTH) {
-        if (cert_list->next != NULL || cert_list->prev != NULL) {
-            DEBUG(SSSDBG_FATAL_FAILURE,
-                  "More than one certificate found for authentication, "
-                  "aborting!\n");
-            ret = EINVAL;
-            goto done;
-        }
-
-        ret = sign_data(module, session, slot_id, cert_list);
-        if (ret != EOK) {
-            DEBUG(SSSDBG_OP_FAILURE, "sign_data failed.\n");
-            ret = EACCES;
-            goto done;
-        }
-
-        DEBUG(SSSDBG_TRACE_ALL,
-              "Certificate verified and validated.\n");
-    }
-
-    multi = talloc_strdup(mem_ctx, "");
-    if (multi == NULL) {
-        DEBUG(SSSDBG_CRIT_FAILURE, "Failed to create output string.\n");
-        ret = ENOMEM;
-        goto done;
-    }
-
-    DLIST_FOR_EACH(item, cert_list) {
-        DEBUG(SSSDBG_TRACE_ALL, "Found certificate has key id [%s].\n",
-              item->id);
 
-        multi = talloc_asprintf_append(multi, "%s\n%s\n%s\n%s\n%s\n",
-                                       token_name, module_file_name, item->id,
-                                       item->label, item->cert_b64);
+        ret = do_slot(module, module_id, slot_id, info, token_info, module_info,
+                      mem_ctx, p11_ctx, mode, pin, module_name_in, token_name_in,
+                      key_id_in, label_in, uri_str, _multi);
     }
```

Comment:
Maybe it would be more readable to condition this block on something like '_multi is not empty' instead of `if (slot_id == (CK_SLOT_ID)-1)`?